### PR TITLE
Fix spoofing and refactor handling of multiple source files

### LINKED ISSUE #1117: StreamData spoofing flag is set too aggressively with multiple source or multiple forecast steps
### What happened?

Currently, the spoofing flag is set disregarding the [loop over potentially multiple source files](https://github.com/ecmwf/WeatherGenerator/blob/638ca1603afd79d26b12a7a8333ca52e1c130143/src/weathergen/datasets/multi_stream_data_sampler.py#L339) (as used with SEVIRI) and [over multiple forecast steps](https://github.com/ecmwf/WeatherGenerator/blob/638ca1603afd79d26b12a7a8333ca52e1c130143/src/weathergen/datasets/multi_stream_data_sampler.py#L373). This leads to incorrect behaviour when:
- There are multiple source files but the last one does not overlap the current sample window (in which case [this](https://github.com/ecmwf/WeatherGenerator/blob/638ca1603afd79d26b12a7a8333ca52e1c130143/src/weathergen/datasets/multi_stream_data_sampler.py#L368) is set)
- Training has multiple forecast steps and the last one is empty 

Also, there should be no dependence of `stream_data.is_spoof` on source since the [problem with FSDP2](https://github.com/pytorch/pytorch/issues/158719) requires gradients to be present, but no source is needed for this (as is the case when a stream is trained diagnostically).

### What are the steps to reproduce the bug?

_No response_

### Hedgedoc link to logs and more information. This ticket is public, do not attach files directly.

_No response_

### PR DESCRIPTION
**Pull request by clessig**
## Description

Fix spoofing and refactor handling of multiple source files

## Issue Number

Closes #1117 

Is this PR a draft? Mark it as draft.

## Checklist before asking for review

-   [ ] I have performed a self-review of my code
-   [ ] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [ ] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

**Comment by clessig** [comment]
Convergence plots:

i2xrej9f_train_val_stream ERA5Copy loss_mse loss_avg_stream ERA5Copy loss_mse loss_avg_stream ERA5 loss_mse loss_avg_

j8t9v2ql_train_val_stream NPPATMS loss_mse loss_avg_stream ERA5 loss_mse loss_avg_stream SurfaceCombined loss_mse loss_avg_

File: packages/common/src/weathergen/common/io.py [hunk 1]
```diff
@@ -47,6 +46,22 @@ class IOReaderData:
     data: NDArray[DType]
     datetimes: NDArray[NPDT64]
 
+    def is_empty(self):
+        """
+        Test if data object is empty
+        """
+        return len(self.data) == 0
+
+    def to_torch(self):
```
Comment (tjhunter) [inline packages/common/src/weathergen/common/io.py]: It makes it messy because this class is supposed to hold numpy data (nad the typechecker rightfully complains about these lines)

You only use that in `.tokenizer.batchify_target` and sources => maybe do it there?

File: packages/common/src/weathergen/common/io.py [hunk 2]
```diff
@@ -68,43 +83,33 @@ def create(cls, other: typing.Any) -> "IOReaderData":
...
-        dx = 0.5
-        dy = 0.5
-        other_copy = deepcopy(other)
-        num_healpix_cells = 12 * 4**hl
-        lons, lats = hp.healpix_to_lonlat(
-            np.arange(0, num_healpix_cells), 2**hl, dx=dx, dy=dy, order="nested"
-        )
-        other_copy.coords = np.stack([lats.deg, lons.deg], axis=-1, dtype=np.float32)
-        other_copy.geoinfos = np.zeros((other_copy.coords.shape[0], geoinfo_size), dtype=np.float32)
+        assert len(others) > 0
```
Comment (tjhunter) [inline packages/common/src/weathergen/common/io.py]: `, len(others)`

File: packages/common/src/weathergen/common/io.py [hunk 3]
```diff
@@ -68,43 +83,33 @@ def create(cls, other: typing.Any) -> "IOReaderData":
         return cls(**dataclasses.asdict(other))
 
     @classmethod
-    def spoof(
-        cls, other: typing.Any, nchannels, datetime, geoinfo_size, mean_of_data
-    ) -> typing.Self:
+    def combine(cls, others: typing.Any) -> "IOReaderData":
```
Comment (tjhunter) [inline packages/common/src/weathergen/common/io.py]: you can set `others: "list[IOReaderData]"` for the hints

I find this whole class  method always weird. I personally prefer outside of the class. 

```
def combine(others: list[IOReaderData]) -> IOReaderData:
   ...
   return IOReaderData(**dataclasses.asdict(other))
```

File: src/weathergen/datasets/data_reader_base.py [hunk 4]
```diff
@@ -230,6 +230,25 @@ def len(self):
         """
         return len(self.data)
```
Comment (tjhunter) [inline src/weathergen/datasets/data_reader_base.py:232]: def remove_nan_coords(self) -> "ReaderData":

File: src/weathergen/datasets/data_reader_base.py [hunk 5]
```diff
@@ -230,6 +230,25 @@ def len(self):
...
+        Remove all data points where coords are NaN
+
+        Returns
+        -------
+        self
+        """
+        idx_valid = ~np.isnan(self.coords)
+        # filter should be if any (of the two) coords is NaN
+        idx_valid = np.logical_and(idx_valid[:, 0], idx_valid[:, 1])
+        # apply
```
Comment (tjhunter) [inline src/weathergen/datasets/data_reader_base.py]: would you mind not mutating the data? it makes no difference for perf, it is shorter in code and your usage already implies it could be ref. transparent:

```
        return ReaderData(
            coords=self.coords[idx_valid],
            geoinfos=self.geoinfos[idx_valid],
            data=self.data[idx_valid],  
            datetimes=self.datetimes[idx_valid],
        )
```

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 6]
```diff
@@ -42,6 +41,32 @@
 logger = logging.getLogger(__name__)
 
 
+def collect_sources(stream_ds: list, idx: int, type: str) -> IOReaderData:
```
Comment (tjhunter) [inline src/weathergen/datasets/multi_stream_data_sampler.py]: `list[DataReaderBase]`

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 7]
```diff
@@ -42,6 +41,32 @@
...
+
+    for ds in stream_ds:
+        if type == "source":
+            get_reader_data = ds.get_source
+            normalize_channels = ds.normalize_source_channels
+        elif type == "target":
+            get_reader_data = ds.get_target
+            normalize_channels = ds.normalize_target_channels
+        else:
+            assert False, "invalid type"
```
Comment (tjhunter) [inline src/weathergen/datasets/multi_stream_data_sampler.py]: we should make the assertions informative, otherwise the only way to have relevant information is to make a new run: `assert False, type` or `assert type in ["source", "target"], type`

File: src/weathergen/datasets/stream_data.py [hunk 8]
```diff
@@ -250,123 +263,13 @@ def empty(self):
 
         return self.source_empty() and self.target_empty()
 
-    ####################################################################################################
-    def _merge_cells(self, s_list: list, num_healpix_cells: int, arr_type=torch.Tensor) -> list:
```
Comment (tjhunter) [inline src/weathergen/datasets/stream_data.py:254]: thank you for all the cleanup in that class, much appreciated!

File: src/weathergen/datasets/tokenizer_forecast.py [hunk 9]
```diff
@@ -88,31 +84,27 @@ def batchify_target(
         self,
         stream_info: dict,
         sampling_rate_target: float,
-        coords: np.array,
-        geoinfos: np.array,
-        source: np.array,
-        times: np.array,
+        rdata: IOReaderData,
         time_win: tuple,
-        normalizer,  # dataset
```
Comment (tjhunter) [inline src/weathergen/datasets/tokenizer_forecast.py:96]: since you are mutating rdata as a side effect, I would make a clone:

```
from dataclasses import replace
....
       def batch...(...):
        rdata = replace(rdata)
```

File: src/weathergen/train/loss_calculator.py [hunk 10]
```diff
@@ -310,14 +312,6 @@ def compute_loss(
...
-            # streams_data[i] are samples in batch
-            # streams_data[i][0] is stream 0 (sample_idx is identical for all streams per sample)
-            _logger.warning(
-                f"Loss is 0.0 for sample(s): {[sd[0].sample_idx.item() for sd in streams_data]}."
-                + "This will likely lead to errors in the optimization step."
-            )
-
         # normalize by all targets and forecast steps that were non-empty
         # (with each having an expected loss of 1 for an uninitalized neural net)
         loss = loss / ctr_streams
```
Comment (shmh40) [inline src/weathergen/train/loss_calculator.py:317]: Could we get division by 0 here if stream is true and hence ctr_streams could be = 0?

File: packages/common/src/weathergen/common/io.py [hunk 11]
```diff
@@ -68,43 +83,33 @@ def create(cls, other: typing.Any) -> "IOReaderData":
...
-        assert other_copy.datetimes.shape[0] == n_datapoints, (
-            "number of datapoints do not match data"
-        )
+            coords = np.concatenate([coords, other.coords])
+            geoinfos = np.concatenate([geoinfos, other.geoinfos])
+            data = np.concatenate([data, other.data])
+            datetimes = np.concatenate([datetimes, other.datetimes])
 
-        return cls(**dataclasses.asdict(other_copy))
+        return cls(**dataclasses.asdict(other))
```
Comment (shmh40) [inline packages/common/src/weathergen/common/io.py]: As discussed, should be coords, geoinfos etc. with appropriate syntax not other

File: src/weathergen/datasets/multi_stream_data_sampler.py (hunk 6 above)
Comment (shmh40) [inline src/weathergen/datasets/multi_stream_data_sampler.py]: Rename to collect_data_sources or collect_separate_data_files or similar

File: src/weathergen/datasets/multi_stream_data_sampler.py [hunk 12]
```diff
@@ -42,6 +41,32 @@
...
 
 
+def collect_sources(stream_ds: list, idx: int, type: str) -> IOReaderData:
+    """
+    Utility function to collect all sources / targets from streams list
+    """
+
+    rdatas = []
+
+    for ds in stream_ds:
```
Comment (shmh40) [inline src/weathergen/datasets/multi_stream_data_sampler.py]: stream_datasources perhaps

File: src/weathergen/datasets/multi_stream_data_sampler.py (hunk 7 above)
Comment (clessig) [reply src/weathergen/datasets/multi_stream_data_sampler.py]: Changed. But the only way to really understand what is going wrong is to run again with `loader_num_workers=0`.

**Review by shmh40 — APPROVED** [review]
